tests/test_pkg.py

```diff
@@ -1,35 +1,107 @@
+"""Check the content of source code and test build source distribution.
+
+Binary distribution (wheel) is checked in test workflow.
+"""
+
 from __future__ import annotations
 
-import os
+import subprocess
+import tarfile
 from glob import glob
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+from monty.tempfile import ScratchDir
+
+if TYPE_CHECKING:
+    from pymatgen.util.typing import PathLike
+
+PROJECT_ROOT = Path(__file__).parent.parent
+NAMESPACE_PKGS = {"analysis", "ext", "io"}  # TODO: double check
+
+
+def test_source_code():
+    """Check the source code in the working directory."""
+    src_txt_path = PROJECT_ROOT / "src/pymatgen.egg-info/SOURCES.txt"
+
+    _check_src_txt_is_complete(PROJECT_ROOT, src_txt_path)
+
+    # Check existence of `py.typed` file in each sub-package
+    _check_py_typed_files(PROJECT_ROOT / "src/pymatgen")
+
+
+def test_build_source_distribution():
+    """Test build the source distribution (sdist), also check `py.typed` files."""
+    with ScratchDir("."):
+        # Build the source distribution
+        subprocess.run(["python", "-m", "pip", "install", "--upgrade", "build"], check=True)
+        subprocess.run(["python", "-m", "build", "--sdist", PROJECT_ROOT, "--outdir", ".", "-C--quiet"], check=True)
 
-import pytest
+        # Decompress sdist
+        sdist_file = next(Path(".").glob("*.tar.gz"))
+        sdist_dir = Path(sdist_file.name.removesuffix(".tar.gz"))
+        with tarfile.open(sdist_file, "r:gz") as tar:
+            # TODO: remove attr check after only 3.12+
+            if hasattr(tarfile, "data_filter"):
+                tar.extractall("", filter="data")
+            else:
+                tar.extractall("")  # noqa: S202
 
-SRC_TXT_PATH = "src/pymatgen.egg-info/SOURCES.txt"
+        # Check existence of `py.typed` file in each sub-package
+        _check_py_typed_files(sdist_dir / "src/pymatgen")
 
 
-@pytest.mark.skipif(
```

Comment:
I have a feeling that making `test_egg_sources_txt_is_complete` optional is not a good idea, we should setup the test such that it would always run in unit test, otherwise we may or may not be checking the contents at all:
https://github.com/materialsproject/pymatgen/blob/31f1e1fb8cc1bb517d59ea8e484965bb641c42a0/tests/test_pkg.py#L11-L15

I would update `test_pkg` to cover bdist/sdist build